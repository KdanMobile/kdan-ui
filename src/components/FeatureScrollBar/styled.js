import styled, { css } from 'styled-components';
import { layout, space } from 'styled-system';
import propTypes from '@styled-system/prop-types';
import { colors, size } from '../../themes';

export const BarOuter = styled.div`
  width: 6px;
  height: ${size.bar_height}px;
  border: 0 solid rgba(0, 0, 0, 0.1);
  border-radius: 10px;
  background: rgba(0, 0, 0, 0.1);

  ${({ isFixed, top, left, right }) =>
    isFixed &&
    css`
      position: fixed;
      top: ${top};
      left: ${left};
      right: ${right};
    `}

  ${layout}
  ${space}
`;

export const Bar = styled.div`
  width: 6px;
  height: ${({ value }) => value + '%'};
  transition: height 0.8s;
  border: 0 solid ${colors.white};
  border-radius: 10px;
  background: ${colors.white};
`;

BarOuter.prototype = {
  ...propTypes.layout,
  ...propTypes.space
};
