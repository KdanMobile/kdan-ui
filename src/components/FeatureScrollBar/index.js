import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { BarOuter, Bar } from './styled';

const FeatureScrollBar = ({ max, now, isFixed, top, left, right }) => {
  const [height, setHeight] = useState(0);
  useEffect(() => {
    if (now <= max && now >= 0) {
      setHeight((now / max) * 100);
    }
  }, [now]);

  return (
    <BarOuter isFixed={isFixed} top={top} left={left} right={right}>
      <Bar data-testid='bar' barHeight={height} />
    </BarOuter>
  );
};

FeatureScrollBar.defaultProps = {
  max: 3,
  now: 1,
  isFixed: false,
  top: '0px',
  left: 'unset',
  right: 'unset'
};

FeatureScrollBar.propTypes = {
  /**
    scroll bar 最大數值
  */
  max: PropTypes.number,
  /**
    scroll bar 現在目前數值
  */
  now: PropTypes.number,
  /**
    是否開啟 position:fixed 設定
  */
  isFixed: PropTypes.bool,
  /**
    如果 isFixed 是 true 會設定 css top 屬性。
   */
  top: PropTypes.string,
  /**
    如果 isFixed 是 true 會設定 css left 屬性。
   */
  left: PropTypes.string,
  /**
    如果 isFixed 是 true 會設定 css right 屬性。
   */
  right: PropTypes.string
};

export default FeatureScrollBar;
