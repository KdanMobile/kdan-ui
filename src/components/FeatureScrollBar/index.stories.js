import React from 'react';

import { Container } from '../Decorator';
import FeatureScrollBar from './';

export default {
  component: FeatureScrollBar,
  title: 'FeatureScrollBar',
  decorators: [(story) => <Container styleType='middle'>{story()}</Container>],
  parameters: {
    docs: {
      description: {
        component: 'FeatureBanner 裡面需要用到的 scroll bar。'
      }
    }
  }
};

export const Base = (args) => <FeatureScrollBar {...args} />;
Base.args = {
  isFixed: false,
  max: 3,
  now: 1,
  top: '0px',
  left: '0px'
};
