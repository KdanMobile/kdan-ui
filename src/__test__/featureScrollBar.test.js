import React from 'react';
import { render, cleanup } from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom';

import FeatureScrollBar from '../components/FeatureScrollBar';

describe('FeatureScrollBar', () => {
  afterEach(cleanup);

  test('basic', () => {
    const { container } = render(<FeatureScrollBar />);
    expect(container).toMatchSnapshot();
  });
});
