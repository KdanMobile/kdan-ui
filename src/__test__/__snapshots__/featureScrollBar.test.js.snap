// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureScrollBar basic 1`] = `
.c0 {
  width: 6px;
  height: 168px;
  border: 0 solid rgba(0,0,0,0.1);
  border-radius: 10px;
  background: rgba(0,0,0,0.1);
}

.c1 {
  width: 6px;
  height: 33.33333333333333%;
  -webkit-transition: height 0.8s;
  transition: height 0.8s;
  border: 0 solid #ffffff;
  border-radius: 10px;
  background: #ffffff;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      value="33.33333333333333"
    />
  </div>
</div>
`;
